### DISCUSSION TITLE
xtnsis odd behavior (could be user error)

### DISCUSSION THREAD

**Discussion by utkonos** [binref/refinery#61] [Q&A] [answered]
I may be doing this all wrong. These commands end with `failure in xtnsis: exception of type ValueError; Unable to find an NSIS archive marker.`

<img width="1391" alt="image" src="https://github.com/user-attachments/assets/6d333eb2-b835-4316-a44e-84e4793bb333">

If I try `emit 1.exe | xtnsis` it looks like it does some of the extraction and the decompilation, but there's something wrong.

<img width="1377" alt="image" src="https://github.com/user-attachments/assets/e553e04d-1026-40e6-9fc8-dae1ddafd721">

**Comment by jhhcs** [comment, marked as answer]
When you run 
```
emit 1.exe | xtnsis
```
without anything to follow it up, `xtnsis` will simply extract all the files in the archive and write them to STDOUT, separated by newlines. This is a good default for units that produce text, but with an archive extraction tool like `xtnsis` you likely want to proceed differently.

You might want to first run:
```
emit 1.exe | xtnsis -l
```
which will show you a list of the files that can be extracted. Then, assuming you want to dump all of these files to disk, you can run the following command:
```
emit 1.exe | xtnsis [| dump extracted/{path} ]
```
which will write all the files that were previously listed to a subdirectory called `extracted`. If you do not want to dump all of these files, but only some of them, you can specify selector file masks to `xtnsis`. For example, in order to only extract files matching `*.exe`:
```
emit 1.exe | xtnsis '*.exe' [| dump extracted/{path} ]
```
Finally, if you only want the setup script, since that one is printable, you could type:
```
emit 1.exe | xtnsis setup.nsis
```
to extract just the disassembly of that script that and print it to STDOUT.

**Comment by utkonos** [comment]
Great, thanks! I figured it was user error. Good stuff with binary refinery.

**Comment by utkonos** [comment]
One more followup question: I see that setup.nsis is the decompiled representation of the scirpt. Is there a way to have binref output the original bytecode representation of the script/s in addition to the decompilation?

**Reply by huettenhain** [reply]
Not implemented yet, but shouldn't be a hard thing to do. I'll look into it.

**Reply by huettenhain** [reply]
See #62.

**Comment by utkonos** [comment]
Sounds good. Thanks!
